File: JAVA/src/product/ShoppingCart.java
package product;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;
import utility.ToString;

public class ShoppingCart extends Product {
    private static ShoppingCart singleInstance = null;
    private HashMap<String, Integer> cart = new HashMap<>();

    private ShoppingCart() {
    }

    public static ShoppingCart getInstance() {
        if (singleInstance == null) {
            singleInstance = new ShoppingCart();
        }
        return singleInstance;
    }

    @Override
    public void addItem(Item item) {
        this.cart.put(item.getName(), item.getAmount());
    }

    @Override
    public String getItems() {
        return ToString.cartItemToString(cart);
    }

    public int removeItem(String name) {
        int amount = countItemInCart(name);
        if (amount > 0) {
            this.cart.remove(name);
        }
        return amount;
    }

    public List<Item> removeAllItem() {
        ArrayList<Item> items = new ArrayList<>();
        for (Entry<String, Integer> entry : cart.entrySet()) {
            items.add(new Item(entry.getKey(), entry.getValue()));
        }
        cart.clear();
        return items;
    }

    private int countItemInCart(String name) {
        if (this.cart.containsKey(name)) {
            return this.cart.get(name);
        }
        return 0;
    }
}

File: TYPESCRIPT/src/product/ShoppingCart.ts
import { IProduct } from "./IProduct";
import { Item } from "./Item";

export class ShoppingCart implements IProduct {
  private static _singleInstance = null;
  private _cart: Map<string, number> = new Map();

  private constructor() {}

  public static getInstance(): ShoppingCart {
    if (this._singleInstance === null) {
      this._singleInstance = new ShoppingCart();
    }
    return this._singleInstance;
  }

  public addItem(item: Item): void {
    this._cart.set(item.getName, item.getAmount);
  }

  /**
   * TODO  change this after ToString util finished.
   */
  public getItems(): any {
    return this._cart;
  }

  public removeItem(name: string): number {
    let amount: number = this._countItemInCart(name);
    if (amount > 0) {
      this._cart.delete(name);
    }
    return amount;
  }

  public removeAllItem(): Array<Item> {
    let items: Array<Item> = new Array();
    for (let entry of this._cart) {
      items.push(new Item(entry[0], entry[1]));
    }
    this._cart.clear();
    return items;
  }

  private _countItemInCart(name: string): number {
    if (this._cart.has(name)) {
      return this._cart.get(name);
    }
    return 0;
  }
}
